src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientDiagnosticListener.cs

```diff
@@ -56,6 +56,13 @@ public override void OnCustom(string name, Activity activity, object payload)
                 case SqlDataBeforeExecuteCommand:
                 case SqlMicrosoftBeforeExecuteCommand:
                     {
+                        _ = this.commandFetcher.TryFetch(payload, out var command);
+
+                        if (this.options?.EventFilter(command) == false)
```

Comment:
* `options` can't be null here can it?

* If users aren't using the filter (as it is now) they still have to pay for the call. Might be best just to do the work in-line here. Eg:

```csharp
            try
            {
                if (options.Filter?.Invoke(payload) == false)
                {
                   SqlClientInstrumentationEventSource.Log.EventFilteredOut(activity.OperationName); // <- Probably need to log this as well?
                   return;
                }
            }
            catch (Exception ex)
            {
                SqlClientInstrumentationEventSource.Log.EventFilterException(ex);
                return;
            }
```

If you really want to keep it a separate method you might be able to achieve the same perf with `[MethodImpl(MethodImplOptions.AggressiveInlining)]` not sure, would have to benchmark it. I suspect this will be a smidge faster.

PS: I think if it is a standalone method it might be better off in this file rather than in the options class.

Reply:
`Filter` should only be executed if the sampling decision is favorable, i.e., `activity.IsAllDataRequested == true`. You could move the `Filter` logic to some lines below where they check for this condition.

Review the placement of `Filter` logic in Http and AspNetCore instrumentation:
- https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs#L112
- https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L122